### PR TITLE
APIv4 - Setting api misc fixes & tests

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Fixes the setting api to correctly handle pseudoconstants & field suffixes

Before
----------------------------------------
Setting pseudoconstants not handled well.

After
----------------------------------------
Works better & covered by tests.

### CODE REVIEW

File: CRM/Core/SelectValues.php [hunk 1]
```diff
@@ -437,7 +437,7 @@ public static function geoProvider() {
    *
    * @throws \CRM_Core_Exception
    */
-  public function taxDisplayOptions() {
+  public static function taxDisplayOptions() {
```
Comment (colemanw) [inline CRM/Core/SelectValues.php:440]: @eileenmcnaughton the new unit test which checks settings pseudoconstants hit this (PHP error non-static method called statically).

File: tests/phpunit/api/v3/SettingTest.php [hunk 2]
```diff
@@ -111,11 +111,9 @@ public function testGetFieldsFilters($version) {
 
   /**
    * Test that getfields will filter on group.
-   * @param int $version
-   * @dataProvider versionThreeAndFour
    */
-  public function testGetFieldsGroupFilters($version) {
-    $this->_apiversion = $version;
+  public function testGetFieldsGroupFilters() {
+    $this->_apiversion = 3;
```
Comment (colemanw) [inline tests/phpunit/api/v3/SettingTest.php:116]: `group` is a deprecated property so APIv4 doesn't need to test it.

File: Civi/Core/SettingsMetadata.php [hunk 3]
```diff
@@ -69,9 +69,7 @@ public static function getMetadata($filters = [], $domainID = NULL, $loadOptions
     }
 
     self::_filterSettingsSpecification($filters, $settingsMetadata);
-    if ($loadOptions) {
```
Comment (eileenmcnaughton) [inline Civi/Core/SettingsMetadata.php:72]: @colemanw so this seems to result in us iterating through every setting and doing noting if loadOptions is false - as opposed to just skipping. Why the change?

Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php:72]: Also - I note loadOptions is not cached so that doesn't mitigate the empty iteration

File: Civi/Core/SettingsMetadata.php [hunk 4]
```diff
@@ -141,30 +139,54 @@ protected static function _filterSettingsSpecification($filters, &$settingSpec)
    * Retrieve options from settings metadata
    *
    * @param array $settingSpec
+   * @param bool|array $getOptions
    */
-  protected static function loadOptions(&$settingSpec) {
+  protected static function fillOptions(&$settingSpec, $getOptions) {
```
Comment (eileenmcnaughton) [inline Civi/Core/SettingsMetadata.php:144]: So the thing I find kinda messy is that getOptions is really a 3 way option it

- NO
- YES
- do it as an array

I feel like is_array() to determine that last one is a bit opaque

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:144]: Well, yea but that's way outside the scope of this PR IMO.

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:144]: It's really an APIv4 documentation issue.

File: tests/phpunit/api/v3/SettingTest.php (hunk 2 above)
Comment (eileenmcnaughton) [reply tests/phpunit/api/v3/SettingTest.php:116]: Does apiv4 still permit filtering on it if set. We kinda deprecated it as a required key but kept the idea you could filter by it (e.g if you wanted to present all the settings defined by one extension on a page). I can't recall if it was reliable or not

File: Civi/Core/SettingsMetadata.php (hunk 4 above)
Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php:144]: @colemanw well maybe at the api level - but at the level of this function we could pass something pretty clear

File: Civi/Core/SettingsMetadata.php (hunk 3 above)
Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:72]: You're right. I originally had this doing more but had to change it back due to a failing test.

File: Civi/Core/SettingsMetadata.php (hunk 4 above)
Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:144]: Unfortunately, I don't think we can. To maintain backward-compat it needs to be able to return a flat array if `true` and in order to correctly serve data to APIv4 it needs to be able to fetch name, label, etc if array.

File: tests/phpunit/api/v3/SettingTest.php (hunk 2 above)
Comment (colemanw) [reply tests/phpunit/api/v3/SettingTest.php:116]: No. I was following [the docs ](https://docs.civicrm.org/dev/en/latest/framework/setting/#deprecated-properties)which state it is a deprecated property. I didn't want APIv4 returning stuff that's already deprecated. That means it wouldn't be able to filter on it either.

File: Civi/Core/SettingsMetadata.php (hunk 3 above)
Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php:72]: @colemanw is it going to do more again soon? If so I guess it's Ok to merge as is

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:72]: Probably not. I've replaced the conditional as before.

File: Civi/Core/SettingsMetadata.php (hunk 4 above)
Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php:146]: Re backward-compat - it's only called from one place in the code so the contract can change I think - I mean you DID rename it!

Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php:146]: in fact getOptions now should be renamed to $returnOptionsAsArray since it's a new param to the function

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:146]: Well the function is only called from one place but it returns values to `SettingsMetadata::getMetadata` which is used in legacy places that expect flat arrays of options.
Since it's protected I took the liberty of renaming it :)

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php:146]: Sure I'll rename that param too.

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php]: I renamed it but resisted the temptation to turn it into a boolean because eventually we'll add support for more return values.

Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php]: Hmm I feel like Civi is full of that sort of good intention code which makes less & less sense after the extra thing isn't done months later. I'd probably go with what actually makes sense with the code now as it's a protected function - but failing that beef up the comments explaining the rationale for the param in the docblock

Comment (colemanw) [reply Civi/Core/SettingsMetadata.php]: @eileenmcnaughton I renamed the variable & added a comment. I don't think this is one of those cases because it's not broken or half-functional. It works, this just leaves the door open for further enhancements.

Comment (eileenmcnaughton) [reply Civi/Core/SettingsMetadata.php]: OK - it still feels a bit too much like magic to me but at least the docs cover it now
